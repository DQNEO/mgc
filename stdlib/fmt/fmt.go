package fmt

import (
	"os"
	"github.com/DQNEO/minigo/stdlib/reflect"
	"github.com/DQNEO/minigo/stdlib/strconv"
	"github.com/DQNEO/minigo/stdlib/io"
)

func Print(s string) {
	os.Stdout.Write([]byte(s))
}

func Println(s string) {
	b := []byte(s)
	b = append(b, '\n')
	os.Stdout.Write(b)
}

func Fprintf(w io.Writer, format string, a ...interface{}) {
	p := newPrinter()
	p.doPrintf(format, a)
	w.Write(p.buf)
	p.free()
}

func Sprintf(format string, a ...interface{}) string {
	p := newPrinter()
	p.doPrintf(format, a)
	s := string(p.buf)
	p.free()
	return s
}

func Printf(format string, a ...interface{}) {
	Fprintf(os.Stdout, format, a...)
}

type printer struct {
	buf []byte
}

func newPrinter() *printer {
	p := &printer{}
	return p
}

func (p *printer) free() {
	p.buf = p.buf[0:0]
}

func (p *printer) doPrintf(format string, a []interface{}) {

	var buf string
	var tmpbuf []byte
	var inPercent bool
	var argIndex int

	for _, c := range []byte(format) {
		if inPercent {
			if c == '%' {
				tmpbuf = append(tmpbuf, c)
			} else {
				tmpbuf = nil
				arg := a[argIndex]
				argIndex++
				s := p.printArg(arg, c)
				buf = buf + s
			}
			inPercent = false
		} else {
			if  c == '%' {
				inPercent = true
				buf = buf + string(tmpbuf)
				tmpbuf = nil
			} else {
				tmpbuf = append(tmpbuf, c)
			}
		}
	}
	buf = buf + string(tmpbuf)

	p.buf = []byte(buf)
}

func (p *printer) printArg(arg interface{}, verb byte) string {

	switch verb {
	case 'T':
		return reflect.TypeOf(arg).String()
	case 'p':
		return "[==ADDRESS==]"
	}

	var s string
	switch arg.(type) {
	case string: // for %s
		s = arg.(string)
	case []byte: // for %s
		s = string(arg.([]byte))
	case byte: // for %c or %d
		switch verb {
		case 'c':
			b := arg.(byte)
			s = string([]byte{b})
		case 'd':
			b := arg.(byte)
			i := int(b)
			s = strconv.Itoa(i)
		default:
			panic("unknown format flag")
		}
	case int: // for %d
		s = strconv.Itoa(arg.(int))
	case uint16: // for %d
		s = strconv.Itoa(int(arg.(uint16)))
	case bool: // for %v
		if arg.(bool) {
			s = "true"
		} else {
			s = "false"
		}
	default:
		//panic(fmt.Sprintf("%T\n", arg))
		panic("Unkown type to format:")
	}
	return s
}
