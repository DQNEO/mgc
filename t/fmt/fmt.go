package main

import "fmt"

func f1() {
	var i int = 1
	fmt.Printf("%d\n", i)

	var c byte = 'a'
	fmt.Printf("%d\n", c)

	i = int(c)
	fmt.Printf("%d\n", i)
}

func main() {
	f1()
}
