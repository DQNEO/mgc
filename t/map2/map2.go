package main

import "fmt"

func f1() {
	var lmap map[int]int = map[int]int{
		1: 2,
		3: 4,
	}

	for i, v := range lmap {
		fmt.Printf("%d\n", i)
		fmt.Printf("%d\n", v)
	}

	fmt.Printf("%d\n", lmap[1]+3) // 5
	fmt.Printf("%d\n", lmap[3]+2) // 6

	lmap[7] = 8
	fmt.Printf("%d\n", lmap[4]+7) // 7
	fmt.Printf("%d\n", lmap[7])   // 8
}

func f2() {
	keyFoo := "15"
	var lmap map[string]string = map[string]string{
		keyFoo:   "10",
		"17": "11",
	}

	fmt.Printf("9%s\n", lmap["noexists"])
	fmt.Printf("%s\n", lmap["15"]) // 10
	fmt.Printf("%s\n", lmap["17"]) // 11


	fmt.Printf("%d\n", len(lmap) + 10) // 12

	lmap["19"] = "13"
	fmt.Printf("%s\n", lmap["19"]) // 13

	fmt.Printf("%d\n", len(lmap) + 11 ) // 14
	lmap["15"] = "16"
	lmap["17"] = "18"
	lmap["19"] = "20"
	for k, v := range lmap {
		fmt.Printf("%s\n%s\n", k,v) // 15,16,17,18,19,20
	}
}

func f3() {
	var lmap map[int]int = map[int]int{
		1: 2,
		3: 21,
	}
	var ok bool
	var val int
	val, ok = lmap[3]
	fmt.Printf("%d\n", val) // 21
	if ok {
		fmt.Printf("%d\n", 22)
	}

	val, ok = lmap[2]
	if !ok {
		fmt.Printf("%d\n", 23)
	}
	fmt.Printf("%d\n", val + 24) //24
}

var keyFoo2 string = "keyfoo"

func f4() {
	keyFoo := "keyfoo"
	var lmap map[string]string = map[string]string{
		keyFoo:   "26",
		"keybar": "valuebar",
	}

	var ok bool
	var v string
	v, ok = lmap[keyFoo2]
	if ok {
		fmt.Printf("%d\n", 25)
	}
	fmt.Printf("%s\n", v) // 26

	v, ok = lmap["noexits"]
	if !ok {
		fmt.Printf("%d\n", 27)
	}
	fmt.Printf("28%s\n", v) // 28
}

func main() {
	f1()
	f2()
	f3()
	f4()
}
