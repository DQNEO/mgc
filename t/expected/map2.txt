1
2
3
4
5
6
7
8
valuefoo
valuebar
valueadded
keyfoo
valuefoo
keybar
valuebar
keyadded
valueadded
